serde_derive/src/de.rs

```diff
@@ -85,37 +85,63 @@ fn requires_default(attrs: &attr::Field) -> bool {
 }
 
 fn deserialize_body(item: &Item, generics: &syn::Generics) -> Fragment {
-    match item.body {
-        Body::Enum(ref variants) => {
-            deserialize_item_enum(&item.ident, generics, variants, &item.attrs)
+    if let Some(fr_ty) = item.attrs.from_type() {
+        if let Some(fr_as_ty) = item.attrs.from_as_type() {
+            deserialize_from_as(fr_ty, fr_as_ty)
+        } else {
+            deserialize_from(fr_ty)
         }
-        Body::Struct(Style::Struct, ref fields) => {
-            if fields.iter().any(|field| field.ident.is_none()) {
-                panic!("struct has unnamed fields");
+    } else {
+        match item.body {
+            Body::Enum(ref variants) => {
+                deserialize_item_enum(&item.ident, generics, variants, &item.attrs)
             }
+            Body::Struct(Style::Struct, ref fields) => {
+                if fields.iter().any(|field| field.ident.is_none()) {
+                    panic!("struct has unnamed fields");
+                }
+                deserialize_struct(&item.ident,
+                                   None,
+                                   generics,
+                                   fields,
+                                   &item.attrs,
+                                   None)
+            }
+            Body::Struct(Style::Tuple, ref fields) |
+            Body::Struct(Style::Newtype, ref fields) => {
+                if fields.iter().any(|field| field.ident.is_some()) {
+                    panic!("tuple struct has named fields");
+                }
+                deserialize_tuple(&item.ident,
+                                  None,
+                                  generics,
+                                  fields,
+                                  &item.attrs,
+                                  None)
+            }
+            Body::Struct(Style::Unit, _) => deserialize_unit_struct(&item.ident, &item.attrs),
+        }
+    }
+}
 
-            deserialize_struct(&item.ident,
-                               None,
-                               generics,
-                               fields,
-                               &item.attrs,
-                               None)
+fn deserialize_from(from_type: &syn::Ty) -> Fragment {
+    quote_block!({
```

Comment:
Just use `quote_block { /* ... */ }` as you did in serialize_into, no need for the extra layer of brace.